### PR TITLE
Fixes #19601 - Cleans up formatting and introduces a new s3.py option 'ignore_nonexistent_bucket'

### LINKED ISSUE #19601: S3 module: GetObject needs ListBucket permissions
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

S3 module

##### ANSIBLE VERSION

```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION

Using the S3 module with boto installed. Instance has a AWS IAM Role attached.

##### OS / ENVIRONMENT

Running Amazon Linux AMI 2016.09.0 (HVM), SSD Volume Type

##### SUMMARY
S3 module needs unnecessary permissions

##### STEPS TO REPRODUCE

1. Create a s3 bucket "mybucket"
2. Create an EC2 instance with an attached IAM Role.
3. Attach a policy to the S3 bucket that allows the IAM role only GetObject, but not ListBucket
4. Install ansible + boto on the ec2 instance.

```yaml
---
- hosts: localhost
  remote_user: ec2-user
  tasks:
    - name: download ec2.py from s3
      s3:
        bucket: mybucket
        object: /ec2.py
        dest: /tmp/ec2.py
        mode: get
```

##### EXPECTED RESULTS

File is downloaded from S3.

##### ACTUAL RESULTS

File can not be downloaded from S3. Message "Source bucket can not be found" is echoed instead.

```
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "bucket": "mybucket", 
            "dest": "/tmp/ec2.py", 
            "ec2_url": null, 
            "encrypt": true, 
            "expiry": "600", 
            "headers": null, 
            "marker": null, 
            "max_keys": "1000", 
            "metadata": null, 
            "mode": "get", 
            "object": "/ec2.py", 
            "overwrite": "always", 
            "permission": [
                "private"
            ], 
            "prefix": null, 
            "profile": null, 
            "region": null, 
            "retries": 0, 
            "rgw": false, 
            "s3_url": null, 
            "security_token": null, 
            "src": null, 
            "validate_certs": true, 
            "version": null
        }, 
        "module_name": "s3"
    }, 
    "msg": "Source bucket cannot be found"
}
```
Downloading the file using the AWS CLI tools works fine.
After toying around for a few hours, and reiterating all steps several times with small changes, I found the issue: The S3 module with get only works if the IAM Role has ListBucket permissions on the bucket in addition to the GetObject permissions.

I have:

- Verified my boto version is new enough (boto 2.42)
- Verified the EC2 instance role has the correct permissions (aws s3 cp s3://mybucket/ec2.py /tmp/ec2.py works just fine)
- Verified the EC2 instance credentials are available through curl http://169.254.169.254/latest/meta-data/iam/security-credentials/s3access

The error message is not very helpful, and is actually wrong - the bucket can be found with the AWS S3 api.

Suggestion: Do not require ListBucket permissions ob bucket for GetObject operation, or at least show a correct error message.

### PR DESCRIPTION
**Pull request by s-hertel**
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/s3.py

##### ANSIBLE VERSION
```
ansible 2.3.0
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
Fixes #19601. Improves error messages and cleans up some formatting and excessive code. Also introduces a new option 'ignore_nonexistent_bucket' (the default is False) to appropriately set the 'validate' kwarg for s3.lookup.
Example for the use of ignore_nonexistent_bucket: if a user has the GetObject permission but no other permissions, mode: get will fail since the module by default attempts to lookup the bucket. If ignore_nonexistent_buckets is set to True a user with GetObject permission but no other permissions will be able to get the object.

Before:
```
ansible-playbook s3-test.yml
 [WARNING]: Host file not found: /etc/ansible/hosts

 [WARNING]: provided hosts list is empty, only localhost is available


PLAY [localhost] ****************************************************************************************

TASK [Gathering Facts] **********************************************************************************
 [WARNING]: Removed unexpected internal key in module return: _ansible_verbose_override = True

ok: [localhost]

TASK [download file from s3] ****************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Source bucket cannot be found"}
	to retry, use: --limit @/Users/shertel/Workspace/01-18-17/ansible/s3-test.retry

PLAY RECAP **********************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

After:
```
ansible-playbook s3-test.yml
 [WARNING]: Host file not found: /etc/ansible/hosts

 [WARNING]: provided hosts list is empty, only localhost is available

PLAY [localhost] ****************************************************************************************

TASK [Gathering Facts] **********************************************************************************
 [WARNING]: Removed unexpected internal key in module return: _ansible_verbose_override = True

ok: [localhost]

TASK [download file from s3] ****************************************************************************
changed: [localhost]

PLAY RECAP **********************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/s3.py [hunk 1]
```diff
@@ -554,118 +565,110 @@ def main():
     if s3 is None: # this should never happen
         module.fail_json(msg ='Unknown error, failed to create s3 connection, no information from boto.')
 
+    # First, we check to see if the bucket exists, we get "bucket" returned.
+    bucketrtn = bucket_check(module, s3, bucket)
+    
+    if not ignore_nonexistent_bucket:
+        validate = True
+        if mode not in ('create', 'put', 'delete') and not bucketrtn:
+            module.fail_json(msg="Source bucket cannot be found.", failed=True)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/s3.py]: The `failed=True` argument isn't needed here.

File: lib/ansible/modules/cloud/amazon/s3.py [hunk 2]
```diff
@@ -554,118 +565,110 @@ def main():
...
-
         # Next, we check to see if the key in the bucket exists. If it exists, it also returns key_matches md5sum check.
-        keyrtn = key_check(module, s3, bucket, obj, version=version)
+        keyrtn = key_check(module, s3, bucket, obj, version=version, validate=validate)
         if keyrtn is False:
             if version is not None:
                 module.fail_json(msg="Key %s with version id %s does not exist."% (obj, version), failed=True)
             else:
-                module.fail_json(msg="Key %s does not exist."%obj, failed=True)
+                module.fail_json(msg="Key %s does not exist."%obj, failed=True)  # we could change this to specify that the problem may also be a nonexistent bucket
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/s3.py]: Can you update this message before merging?

File: lib/ansible/modules/cloud/amazon/s3.py [hunk 3]
```diff
@@ -554,118 +565,110 @@ def main():
...
+            download_s3file(module, s3, bucket, obj, dest, retries, version=version, validate=validate)
 
         # Compare the remote MD5 sum of the object with the local dest md5sum, if it already exists.
         if pathrtn is True:
-            md5_remote = keysum(module, s3, bucket, obj, version=version)
+            md5_remote = keysum(module, s3, bucket, obj, version=version, validate=validate)
             md5_local = module.md5(dest)
             if md5_local == md5_remote:
                 sum_matches = True
+                module.exit_json(msg="Local and remote object are identical, ignoring. Use overwrite=always parameter to force.", changed=False)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/s3.py]: This will cause the `overwrite=always` setting to never work, since the module exits before checking the value of `overwrite`. Please delete this line.

File: lib/ansible/modules/cloud/amazon/s3.py [hunk 4]
```diff
@@ -150,6 +150,12 @@
     default: null
     aliases: []
     version_added: "1.3"
+  ignore_nonexistent_bucket:
+    description:
+      - "Overrides initial bucket lookups in case bucket or iam policies are restrictive. Example: a user may have the GetObject permission but no other permissions. In this case using the option mode: get will fail without specifying ignore_nonexistent_bucket: True."
+    default: false
+    aliases: []
+    version_added: "2.2"
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/s3.py]: This'll need to be updated to 2.3

**Review by ryansb — CHANGES_REQUESTED** [review]
Overall looks good. I'll test locally as well.

**Comment by s-hertel** [comment]
@lwade

File: lib/ansible/modules/cloud/amazon/s3.py (hunk 3 above)
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/s3.py]: Line 575 checks `overwrite`, but I think that block is redundant. Line 579 `if pathrtn is True:` has an else on line 588 and it also checks `overwrite`.  So as it was if overwrite is always it will attempt to download once on 576 but then a second time below. Should I revert the change I made here and then delete lines 575 and 576?

**Review by ryansb — APPROVED** [review]

File: lib/ansible/modules/cloud/amazon/s3.py [hunk 5]
```diff
@@ -632,107 +647,88 @@ def main():
...
                 dirobj = obj + "/"
-            if bucketrtn is True:
+            if bucketrtn:
                 keyrtn = key_check(module, s3, bucket, dirobj)
                 if keyrtn is True:
                     module.exit_json(msg="Bucket %s and key %s already exists."% (bucket, obj), changed=False)
                 else:
                     create_dirkey(module, s3, bucket, dirobj)
-            if bucketrtn is False:
+            if bucketrtn:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/s3.py]: This should be `else`, looks like the logic got inverted accidentally.

**Review by ryansb — COMMENTED** [review]
One minor tweak, then this is good to go.

**Review by ryansb — APPROVED** [review]
